src/core/strvector.c

```diff
@@ -211,19 +211,29 @@ igraph_error_t igraph_strvector_set_len(igraph_strvector_t *sv, igraph_integer_t
 
 void igraph_strvector_remove_section(
         igraph_strvector_t *sv, igraph_integer_t from, igraph_integer_t to) {
+    igraph_integer_t size = igraph_strvector_size(sv);
     igraph_integer_t i;
 
     IGRAPH_ASSERT(sv != NULL);
     IGRAPH_ASSERT(sv->stor_begin != NULL);
 
-    for (i = from; i < to; i++) {
-        IGRAPH_FREE(sv->stor_begin[i]);
+    if (from < 0) {
+        from = 0;
```

Comment:
I don't like this because it hides mistakes. This should be an `IGRAPH_ASSERT()`. Mistakes should be brought to the surface so that they can be fixed.

Reply:
How about the truncation of `to` to the length of the vector? Should that also become an `IGRAPH_ASSERT()`? Note that the current implementation allows clearing the vector from a given index onwards to the end by supplying a very large number to `to`, which would not be possible if we check `to` rigorously.

Also, how about the case of `from > to`? Technically, that's also a mistake, so should that also become an `IGRAPH_ASSERT()`? Most programming languages with slicing operators happily allow the case of `from > to`, they just simply interpret it as an empty slice.

Reply:
If you think that there is a good use case for one of these, then do not prevent it with an assert.

Personally think that supplying a large number to `to` is sloppy: one should supply the size instead: `strvector_remove_section(strvec, from, strvector_size(strvec))`.

As for `from > to`, my gut feeling is that it's fine to interpret this as empty, but I can't quite explain why, in a rational manner. Probably it's my association to C++ ranges: a typical implementation is often similar to 

```
while (from < to) {
    do_stuff(from);
    from++;
}
```

This naturally leads to an empty range when `from >= to`. But if `from < 0` or `to > size` then we have an out-of-bounds access.

Another point is that, subjectively, I think that there's a higher risk of out-of-bounds accesses than an accidental `from > to`, so catching out-of-bounds cases is more useful than catching `from > to`.

Reply:
Thinking more about this,

 - Allowing from < 0 and to > size is mathematically consistent. We can have any range within (-infty, +infty) and the overlap of this range with the vector will be removed.
 - But what matters here is not mathematical consistency, but usefulness. Is there a realistic use case for from < 0 or to > size? I can't think of one right now, even for convenience. It seems to me that if such values are passed in, the most likely reason is that the programmer made a mistake (there's a bug). Thus, the most useful behaviour is to make them aware of the mistake, just in case the mistake would have other consequences as well.

Reply:
I am still of the opinion that downright crashing the app when we could do better is downright rude; supplying negative indices to `to` and `from` are mistakes but they aren't mistakes that we could not solve on our own inside `remove_section()` (unlike things like null pointers in `stor_begin` that we clearly cannot solve). Since the alternative that I've implemented is mathematically consistent, I'd rather leave it that way. (I am partly inspired here by higher-level languages like Python or JavaScript that do not make a fuss about out-of-bounds indices when slicing an array). So, unless you have strong feelings against forcing `from` and `to` into a sensible range, I'll leave it this way.

Reply:
Code written in higher-level language is bug prone exactly because of this.

My point is that if the programmer made a mistake, a crash from `remove_section()` is very likely not the only consequence. Hiding the mistake does a disservice to the programmer: they will notice the problem later and will have to spend more time debugging it.  A "crash" is not evil. It is simply the way errors show up in C. What's helpful to do is to make it a crash that returns useful information (i.e. an `assert`) instead of a plain segfault.

This is the tradeoff between choosing a strongly typed low-level language and a "sloppy" high-level one. In the former, we need to do a bit more work because every aspect of the program must be specified precisely, but in return we get guarantees that we can rely on. The latter lets us get away with a quick and dirty job. That's useful if I just want to get some data analysis done with no fuss, but quick and dirty work will get you dirty results. Eventually something will go wrong. There is _some_ input for which the program will fail.

Since we are using a strict language, and we are working on the innermost, core part of the library, we should go with the philosophy of strict languages: provide as strong guarantees as possible, and surface errors as fast as possible. The high-risk approach of Python is not suitable for C, especially not for a library that must perform reliably under all conditions.

Reply:
Also, this behaviour extends automatically to `strvector_remove()`. Ignoring out-of-range errors there is even worse.

Reply:
>  A "crash" is not evil. It is simply the way errors show up in C. 

In certain contexts it _is_ evil. I wouldn't want igraph to run on any of my drones at this moment due to the tendency of diverting to fatal errors even when it would be perfectly fine to let the user handle the error gracefully. For instance, I could consider using igraph on a drone as a path planner on the visibility graph of a fixed set of hard-coded obstacles, but if I cannot trust igraph not to throw fatal errors, I would not use it. If igraph simply returned an error code, I could just make it a policy not to ignore the return values of igraph functions (there are static code checkers to ensure that), and handle all errors while keeping the basic flight stabilization routines of the firmware operational at all times. It's not like scientific computing where the worst thing that can happen is that you get thrown back to the shell and you need to restart the calculation that you were working on (although this can also be pretty annoying if you submit a few thousand jobs to a cluster and they all crash after a few hours of number crunching because of some internal error in igraph that we decided to make fatal even if we could have handled it).

So, basically, I'd like to leave fatal errors for situations when something is so utterly broken beyond all repair that we cannot do anything sensible about it. Otherwise we could simply take all cases when we return `IGRAPH_EINVAL` into fatal errors, couldn't we?

Anyhow, I think this discussion has far-fetching implications for the entire coding style of the library and it's probably not a PR where we should discuss this in detail. If you wish, we can put this on the agenda of the next dev meeting where all the team members can voice their opinions.

Reply:
If the function received an out-of-bounds value when it was not supposed to, then how can you trust anything your program does at this point? Something went wrong, it's just hidden. Your drone is going to fly down when it's supposed to go up, and you won't even know why.

igraph will only throw fatal errors when there is a bug. If you don't want it to throw fatal errors, you should find the bugs and fix them. Hiding the failures is an anti-pattern. It leads to bugs lurking for longer before they are found and fixed.

BTW fatal errors _can_ be handled in those special situations where it's necessary to be able to recover even from a disaster. But at least we know that there was a disaster.

Reply:
> Otherwise we could simply take all cases when we return `IGRAPH_EINVAL` into fatal errors, couldn't we?

No, as I said last time this is part of the standard argument checking, which high-level interfaces need to do anyway. Invalid input to those functions which are exposed in high-level interfaces is something we expect, and handle explicitly. In other words, if the someone passes in a negative probability from R, that's a bug in the user's R code, not a bug in the C core. That's why it's not a fatal error.